src/crypto/aead/aes256gcm.rs

```diff
@@ -0,0 +1,234 @@
+        crypto_aead_aes256gcm_decrypt_detached, crypto_aead_aes256gcm_encrypt,
+        crypto_aead_aes256gcm_encrypt_detached,
+    };
+
+    /// `is_available` returns true if the current CPU supports aes256gcm and false otherwise.
+    ///
+    /// WARNING: You must call [init](crate::init) before calling this function; if you do not
+    /// `is_available` will always return false even if the runtime supports aes256gcm.
+    pub fn is_available() -> bool {
+        unsafe { ffi::crypto_aead_aes256gcm_is_available() == 1 }
+    }
+
+    aead_module!(
+        crypto_aead_aes256gcm_encrypt,
+        crypto_aead_aes256gcm_decrypt,
+        crypto_aead_aes256gcm_encrypt_detached,
+        crypto_aead_aes256gcm_decrypt_detached,
+        crypto_aead_aes256gcm_KEYBYTES as usize,
+        crypto_aead_aes256gcm_NPUBBYTES as usize,
+        crypto_aead_aes256gcm_ABYTES as usize,
+        crate::init().is_ok() && is_available()
+    );
+
+    #[cfg(test)]
+    mod test {
+        use super::*;
+
+        #[test]
+        fn test_vector_1() {
+            // Test vector from https://tools.ietf.org/html/rfc7714#section-16.2.2
+            let m = &[
+                0x47, 0x61, 0x6c, 0x6c, 0x69, 0x61, 0x20, 0x65, 0x73, 0x74, 0x20, 0x6f, 0x6d, 0x6e,
+                0x69, 0x73, 0x20, 0x64, 0x69, 0x76, 0x69, 0x73, 0x61, 0x20, 0x69, 0x6e, 0x20, 0x70,
+                0x61, 0x72, 0x74, 0x65, 0x73, 0x20, 0x74, 0x72, 0x65, 0x73,
+            ];
+            let ad = &[
+                0x80, 0x40, 0xf1, 0x7b, 0x80, 0x41, 0xf8, 0xd3, 0x55, 0x01, 0xa0, 0xb2,
+            ];
+            let k = Key([
+                0x00, 0x01, 0x02, 0x03, 0x04, 0x05, 0x06, 0x07, 0x08, 0x09, 0x0a, 0x0b, 0x0c, 0x0d,
+                0x0e, 0x0f, 0x10, 0x11, 0x12, 0x13, 0x14, 0x15, 0x16, 0x17, 0x18, 0x19, 0x1a, 0x1b,
+                0x1c, 0x1d, 0x1e, 0x1f,
+            ]);
+            let n = Nonce([
+                0x51, 0x75, 0x3c, 0x65, 0x80, 0xc2, 0x72, 0x6f, 0x20, 0x71, 0x84, 0x14,
+            ]);
+            let c_expected = &[
+                0x32, 0xb1, 0xde, 0x78, 0xa8, 0x22, 0xfe, 0x12, 0xef, 0x9f, 0x78, 0xfa, 0x33, 0x2e,
+                0x33, 0xaa, 0xb1, 0x80, 0x12, 0x38, 0x9a, 0x58, 0xe2, 0xf3, 0xb5, 0x0b, 0x2a, 0x02,
+                0x76, 0xff, 0xae, 0x0f, 0x1b, 0xa6, 0x37, 0x99, 0xb8, 0x7b, 0x7a, 0xa3, 0xdb, 0x36,
+                0xdf, 0xff, 0xd6, 0xb0, 0xf9, 0xbb, 0x78, 0x78, 0xd7, 0xa7, 0x6c, 0x13,
+            ];
+            let c = seal(m, Some(ad), &n, &k);
+            assert_eq!(&c[..].len(), &c_expected[..].len());
+            assert_eq!(&c[0..44], &c_expected[0..44]);
+        }
+    }
+}
+
+mod aes_api {
+    use super::aes_impl;
+    use crypto::nonce::gen_random_nonce;
+
+    /// The Aes256Gcm struct encapsulates the crypto_aead_aes256gcm_* family of
+    /// functions in a way that ensures safe usage of the API at runtime
+    /// without incurring a per function call cost.
+    pub struct Aes256Gcm;
```

Comment:
I think we should derive `Debug`, `Clone` and `Copy` here to make it as easy as possible to embed and pass around this token.